src/storage.ts

```diff
@@ -114,6 +91,23 @@ export class Storage {
       binName += ".exe"
     }
     const controller = new AbortController()
+
+    if (exists) {
+      this.output.appendLine(`Checking if binary outdated...`)
+      const outdated = await this.checkBinaryOutdated(binName, baseURL, controller)
+      // If it's outdated, we fall through to the download logic.
+      if (outdated) {
+        this.output.appendLine(`Found outdated version.`)
+      } else {
+        // Even if the file exists, it could be corrupted.
+        // We run `coder version` to ensure the binary can be executed.
+        this.output.appendLine(`Using existing binary: ${binPath}`)
+        const valid = await this.checkBinaryValid(binPath)
+        if (valid) {
+          return binPath
+        }
+      }
+    }
```

Comment:
I think this logic should be more like:
```
if (exists) {
  const valid = checkBinValid();
  if (valid) {
    return binPath;
  }
  deleteBin();
}

downloadBinary()
```

and the request to the bin endpoint should only be made once. The axios request should be piped to a temp file in the same dir, then once the request is finished:
- if 200: move the temp file on top of the binPath
- if 304: delete the temp file and use the old binPath

Reply:
Using two requests for this is not necessary

Reply:
But doesn't that mean we're always downloading a new binary even if we don't have to?

Reply:
Also if the logic is like that, then we'd never get to the `downloadBinary()` right?

Reply:
I'll see if I can clean up to make it only one request.

Reply:
oh you're right, I meant this:
```
if (exists) {
  const valid = checkBinValid();
  if (!valid) {
    deleteBin();
  }
}

downloadBinary()
```

Reply:
OH, okay that makes more sense. Thank you!